crates/common/lean_chain/src/staker.rs

```diff
@@ -0,0 +1,324 @@
+
+pub struct Staker {
+    pub validator_id: u64,
+    pub chain: HashMap<B256, Block>,
+    pub network: Arc<Mutex<NetworkService>>,
+    pub post_states: HashMap<B256, LeanState>,
+    pub known_votes: Vec<Vote>,
+    pub new_votes: Vec<Vote>,
+    pub dependencies: HashMap<B256, Vec<QueueItem>>,
+    pub genesis_hash: B256,
+    pub num_validators: u64,
+    pub safe_target: B256,
+    pub head: B256,
+}
+
+impl Staker {
+    pub fn new(
+        validator_id: u64,
+        network: Arc<Mutex<NetworkService>>,
+        genesis_block: Block,
+        genesis_state: LeanState,
+    ) -> Staker {
+        let genesis_hash = genesis_block.tree_hash_root();
+
+        Staker {
+            // This node's validator ID
+            validator_id,
+            // Hook to the p2p network
+            network,
+            // Votes that we have received and taken into account
+            known_votes: Vec::new(),
+            // Votes that we have received but not yet taken into account
+            new_votes: Vec::new(),
+            // Objects that we will process once we have processed their parents
+            dependencies: HashMap::new(),
+            // Initialize the chain with the genesis block
+            genesis_hash,
+            num_validators: genesis_state.config.num_validators,
+            // Block that it is safe to use to vote as the target
+            // Diverge from Python implementation: Use genesis hash instead of `None`
+            safe_target: genesis_hash,
+            // Head of the chain
+            head: genesis_hash,
+            // {block_hash: block} for all blocks that we know about
+            chain: HashMap::from([(genesis_hash, genesis_block)]),
+            // {block_hash: post_state} for all blocks that we know about
+            post_states: HashMap::from([(genesis_hash, genesis_state)]),
+        }
+    }
+
+    pub fn latest_justified_hash(&self) -> Option<B256> {
+        get_latest_justified_hash(&self.post_states)
+    }
+
+    pub fn latest_finalized_hash(&self) -> Option<B256> {
+        self.post_states
+            .get(&self.head)
+            .map(|state| state.latest_finalized_hash)
+    }
+
+    /// Compute the latest block that the staker is allowed to choose as the target
+    fn compute_safe_target(&self) -> B256 {
+        let justified_hash = get_latest_justified_hash(&self.post_states).unwrap();
+
+        get_fork_choice_head(
+            &self.chain,
+            &justified_hash,
+            &self.new_votes,
+            self.num_validators * 2 / 3,
+        )
+    }
+
+    /// Process new votes that the staker has received. Vote processing is done
+    /// at a particular time, because of safe target and view merge rule
+    fn accept_new_votes(&mut self) {
+        for new_vote in &self.new_votes {
+            if !self.known_votes.contains(new_vote) {
+                self.known_votes.push(new_vote.clone());
+            }
+        }
+
+        self.new_votes = Vec::new();
```

Comment:
shouldn't we consume `new_votes` instead of cloning its items, then reallocating it at the end? This seems a little wasteful